test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -109,283 +109,245 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+				framework.ExpectNoError(err)
+
+				cleanupSteps = append(cleanupSteps, func() {
+					// The purpose of this block is to prevent physical snapshotContent leaks.
+					// We must update the SnapshotContent to have Delete Deletion policy,
+					// or else the physical snapshot content will be leaked.
+					ginkgo.By("get the latest copy of volume snapshot content")
+					snapshotContent, err := dc.Resource(SnapshotContentGVR).Get(context.TODO(), vscontent.GetName(), metav1.GetOptions{})
+					framework.ExpectNoError(err)
+
+					ginkgo.By("updating the the SnapshotContent to have Delete Deletion policy")
+					snapshotContent.Object["spec"].(map[string]interface{})["deletionPolicy"] = "Delete"
+					vscontent, err = dc.Resource(SnapshotContentGVR).Update(context.TODO(), snapshotContent, metav1.UpdateOptions{})
+					framework.ExpectNoError(err)
+
+					ginkgo.By("manually deleting the SnapshotContent")
+					err = dc.Resource(SnapshotContentGVR).Delete(context.TODO(), snapshotContent.GetName(), metav1.DeleteOptions{})
+					if err != nil && !apierrors.IsNotFound(err) {
+						framework.Failf("Error deleting snapshot content %q. Error: %v", snapshotContent.GetName(), err)
+					}
+
+					ginkgo.By("checking the SnapshotContent has been deleted")
+					err = utils.WaitForGVRDeletion(dc, SnapshotContentGVR, vscontent.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
+					framework.ExpectNoError(err)
+				})
+
+				// Validate volume snapshot content is not removed
+				DeleteAndWaitSnapshot(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
+				err = utils.WaitForGVRDeletion(dc, SnapshotContentGVR, vscontent.GetName(), 1*time.Second /* poll */, 30*time.Second /* timeout */)
+				framework.ExpectError(err) // should fail deletion check
+			})
+
+			ginkgo.It("should create objects with default", func() {
+				ginkgo.By("checking the snapshot")
+				// Get new copy of the snapshot
+				vs, err = dc.Resource(SnapshotGVR).Namespace(vs.GetNamespace()).Get(context.TODO(), vs.GetName(), metav1.GetOptions{})
+				framework.ExpectNoError(err)
+
+				// Get the bound snapshotContent
+				snapshotStatus := vs.Object["status"].(map[string]interface{})
+				snapshotContentName := snapshotStatus["boundVolumeSnapshotContentName"].(string)
+				vscontent, err = dc.Resource(SnapshotContentGVR).Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+
+				snapshotContentSpec := vscontent.Object["spec"].(map[string]interface{})
+				volumeSnapshotRef := snapshotContentSpec["volumeSnapshotRef"].(map[string]interface{})
+
+				// Check SnapshotContent properties
+				ginkgo.By("checking the SnapshotContent")
+				framework.ExpectEqual(snapshotContentSpec["volumeSnapshotClassName"], vsc.GetName())
+				framework.ExpectEqual(volumeSnapshotRef["name"], vs.GetName())
+				framework.ExpectEqual(volumeSnapshotRef["namespace"], vs.GetNamespace())
+			})
```

Comment:
Your test ends up creating and extra snapshot compared to the original test, and it also doesn't verify delete/retain works with whatever is happening in the snapshot check---which, granted, shouldn't matter now, but might if something is added.

I wonder if it would be better to pull out the checks in "should create objects with default" into a common function that is run by the delete & retain steps, or put it into the beforeEach step.

Reply:
Please check the next commit. It's confusing because there was more refactoring done after I had added the new test.

Reply:
Now there is a test pattern which handles "retain" vs "delete" deletion policy. For each It statement it should be testing both policies now. 

Regarding the creating an extra snapshot compared to the original test, I don't believe that is the case, see:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L204

Reply:
From what I understand, in the original test (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go), a snapshot is created in TestSnapshottable, which is run once each on the It()s on lines 189 and 204 when testing delete and retain policies (two snapshots total).

If I'm understanding your refactored code, the delete policy is hoisted up to a testpattern. A snapshot is created in CreateSnapshotResource. That happens before each of the It()s on lines 205 and 218 in https://github.com/kubernetes/kubernetes/blob/9b0dff7686228e2b6f514cdd496e2ae6ab236b9b/test/e2e/storage/testsuites/snapshottable.go. So to test both delete and retain, this code is run through two testpatterns (?) and a total of 4 snapshots are created.

Maybe I've missed something?

Reply:
Take a look at https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L321 and https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/snapshottable.go#L264.

I believe both TestSnapshotDeleted and TestSnapshottable create their own snapshot, and snapshot class. So that would be a total of 4 before and after the change.

Reply:
ah, right you are. Thanks.